toqito/state_props/is_separable.py

```diff
@@ -161,6 +185,79 @@ def is_separable(state: np.ndarray, dim: None | int | list[int] = None, level: i
     #    Xt = state
     # commented out because pylint flagged this as an unused variable
 
+    # There are conditions that are both necessary and sufficient when both
+    # dimensions are 3 and the rank is 4
+    if state_rank == 4 and min_dim == 3 and max_dim == 3:
+        # This method computes more determinants than are actually
+        # necessary, but the speed loss isn't too great
+        p = np.zeros((6, 7, 8, 9))  # initialize.
+        q = orth(state)
+
+        # This loop does not access all positions in `p`, and that's expected.
+        for j in range(6, 0, -1):
+            for k in range(7, j, -1):
+                for n in range(8, k, -1):
+                    for m in range(9, n, -1):
+                        p[j - 1, k - 1, n - 1, m - 1] = np.linalg.det(q[[j - 1, k - 1, n - 1, m - 1], :])
+
+        F = np.linalg.det(
```

Comment:
I don't think `F` is covered by the added unit tests.

Reply:
Yes, it's not covered. From the link you shared, these lines are uncovered:

```
91->98, 129->143, 158, 168-188, 203-269, 292->312, 308, 314-321, 325
```

I could not find any reference providing an example of `state_rank == 4 and min_dim == 3 and max_dim == 3` that reaches that point. A good start for finding one (based on my experiments) might be to look at 9x9 PSD matrices with rank = 4.

Reply:
Hmm, I tried this, but it seemed that it doesn't yield the same values for `j,` `k,` `n,` and `m` as the naive for loop.

The below code is equivalent to the nested for loops:

```py
for j, k, n, m in itertools.product(range(6, 0, -1), range(7, 0, -1), range(8, 0, -1), range(9, 0, -1)):
    if j < k < n < m:
        p[j - 1, k - 1, n - 1, m - 1] = np.linalg.det(q[[j - 1, k - 1, n - 1, m - 1], :])
```

An example:

```py
import itertools
indices1 = []
for j, k, n, m in itertools.product(range(6, 0, -1), range(7, 0, -1), range(8, 0, -1), range(9, 0, -1)):
    if j < k < n < m:
        indices1.append((j,k,n,m)) 

indices2 = []
for j in range(6, 0, -1):
    for k in range(7, j, -1):
        for n in range(8, k, -1):
            for m in range(9, n, -1):
                indices2.append((j,k,n,m))

indices1 == indices2  # Gives True.
```

Reply:
Yeah, let's leave the more verbose one. I think I was trying to be clever and I ended up just proposing something quite silly. Sorry about that!

Reply:
@vprusso, just to clarify, I have used the `itertools` approach I described above since it's equivalent to the for loop. Let me know if you still prefer the more verbose one.

Reply:
Ah, okay. Assuming they are equivalent, either is okay for me. Whichever one you believe to be more clear would be the one I would go with, but I'd be okay with either one.

Reply:
> Another reason to have a test to capture this

I've tried to design the tests a few times, but the challenge I faced was to get 9x9 matrices with rank 4, which is PSD but not PPT, so I could not come up with an example. I think there should be a more principled way to create test cases, but it could take time to implement those. If you have any ideas on how to approach these, I can try them.

Reply:
Yes, this is incredibly specific. I would probably need to set aside some time to dig through the literature to find an appropriate test case matrix for this. If there is an example in the referenced article for this test, there might be something there (and it may be worth giving that a quick skim/check to see if so). 

Otherwise, I'm okay with not having a test case for this condition, but I would like to track this at the very least to ensure that we at least know that this is not explicitly covered.